modules/localization/msf/local_integ/localization_lidar.cc

```diff
@@ -0,0 +1,362 @@
+}
+
+int LocalizationLidar::Update(unsigned int frame_idx,
+                              const Eigen::Affine3d& pose,
+                              const Eigen::Vector3d velocity,
+                              const LidarFrame& lidar_frame) {
+  // check whether loaded map
+  if (_is_map_loaded == false) {
+    _map.LoadMapArea(pose.translation(), _resolution_id,
+                     _zone_id, _search_range_x, _search_range_y);
+    _is_map_loaded = true;
+    LOG(INFO) << "Reflectance locator map first loading is done.";
+  }
+
+  Eigen::Affine3d imu_pose = pose;
+  RefineAltitudeFromMap(&imu_pose);
+
+  // load all needed map
+  Eigen::Vector3d pose_trans = imu_pose.translation();
+  Eigen::Quaterniond pose_quat(imu_pose.linear());
+  pose_quat.normalize();
+  _map.LoadMapArea(pose_trans, _resolution_id,
+                   _zone_id, 0, 0);
+
+  // preload map for next locate
+  _map.PreloadMapArea(pose_trans, velocity, _resolution_id,
+                      _zone_id);
+
+  // generate composed map for compare
+  ComposeMapNode(pose_trans);
+
+  // pass map node to locator
+  int node_width = _lidar_map_node->width;
+  int node_height = _lidar_map_node->height;
+  int node_level_num = 1;
+  _lidar_locator.SetMapNodeData(node_width, node_height, node_level_num,
+      &(_lidar_map_node->intensities), &(_lidar_map_node->intensities_var),
+      &(_lidar_map_node->altitudes), &(_lidar_map_node->count));
+  _lidar_locator.SetMapNodeLeftTopCorner(_map_left_top_corner(0),
+                                         _map_left_top_corner(1));
+
+  // pass lidar points to locator
+  int size = lidar_frame.pt_xs.size();
+  _lidar_locator.SetPointCloudData(
+      size, lidar_frame.pt_xs.data(), lidar_frame.pt_ys.data(),
+      lidar_frame.pt_zs.data(), lidar_frame.intensities.data());
+
+  // compute
+  int error = _lidar_locator.Compute(pose_trans(0),
+                                     pose_trans(1),
+                                     pose_trans(2),
+                                     pose_quat.x(),
+                                     pose_quat.y(),
+                                     pose_quat.z(),
+                                     pose_quat.w());
+
+  return error;
+}
+
+void LocalizationLidar::GetResult(Eigen::Affine3d *location,
+                                  Eigen::Matrix3d *covariance) {
+  double x = 0.0;
+  double y = 0.0;
+  double z = 0.0;
+  double qx = 0.0;
+  double qy = 0.0;
+  double qz = 0.0;
+  double qw = 1.0;
+  _lidar_locator.GetLocationPose(&x, &y, &z, &qx, &qy, &qz, &qw);
+
+  Eigen::Quaterniond quatd(qw, qx, qy, qz);
+  Eigen::Translation3d transd(Eigen::Vector3d(x, y, z));
+  *location = transd * quatd;
+
+  RefineAltitudeFromMap(location);
+
+  int width = 0;
+  int height = 0;
+  const double* data = NULL;
+  _lidar_locator.GetLocationCovariance(&data, &width, &height);
+  for (int row = 0; row < 3; ++row) {
+    for (int col = 0; col < 3; ++col) {
+      (*covariance)(row, col) = data[row * 3 + col];
```

Comment:
How to guarantee row * 3 + col is not out of data's size limit?

Reply:
add check